### ISSUE TITLE
SynchronousFileSink materialized future should fail on error

### ISSUE THREAD

**Issue by pkinsky** [akka/akka#18869]
If a SynchronousFileSink fails due to an upstream error, the future created on materialization completes successfully with the number of bytes written.
```
scala> val failed = Source.failed[ByteString](new Exception("YOLO"))
scala> failed.runWith(SynchronousFileSink(new File("test.txt"))).onComplete(println)
[ERROR] [11/05/2015 15:46:33.604] [default-akka.stream.default-file-io-dispatcher-23]
[akka://default/user/$a/flow-3-1-failedSource-synchronousFileSink] 
Tearing down SynchronousFileSink(/Users/pkinsky/dev/ws-watcher/test.txt) due to upstream error
java.lang.Exception: YOLO
...
Success(0)
```

I think that the correct behavior here is for the future to fail with the throwable that caused the upstream error. This could be accomplished by modifying the code here, in SynchronousFileSubscriber.scala, to include `bytesWrittenPromise.failure(cause)`

```
    case ActorSubscriberMessage.OnError(cause) ⇒
      log.error(cause, "Tearing down SynchronousFileSink({}) due to upstream error", f.getAbsolutePath)
      context.stop(self)
```

However, some applications may require access to the number of bytes written before an upstream failure. If this is the desired behavior, I believe a note should be added to the documentation (if there's not one there already, I couldn't find anything documenting this behavior).

**Comment by pkinsky** [comment]
quick and dirty workaround: 
```
  def safeFileSink(f: File): Sink[ByteString, Future[Long]] = {
    val fileSink = SynchronousFileSink(f)
    val errorMonitor = Sink.ignore
    Sink(fileSink, errorMonitor)((f1, f2) => f2.flatMap(_ => f1)) { implicit builder =>
      (fileSink, errorMonitor) =>
        import FlowGraph.Implicits._

        val broadcast = builder.add(Broadcast[ByteString](2))

        broadcast.out(0) ~> fileSink
        broadcast.out(1) ~> errorMonitor

        broadcast.in
    }
  }
```

**Comment by drewhk** [comment]
Thanks for reporting!

**Comment by rkuhn** [comment]
Hmm, so what shall we do? I do not agree that this version is necessarily more “safe” than the current behavior because safety is subject to the environment in which it is evaluated.

In principle the information that is available and potentially relevant is the combination of how many bytes were written and how the stream ended, so the “truth” is `Sink[ByteString, (Long, Try[Unit])]`, which is quite a mouthful.

**Comment by drewhk** [comment]
> so the “truth” is Sink[ByteString, (Long, Try[Unit])], which is quite a mouthful.

There can be a proper domain class for this, something `IoResult` which contains all the information needed and is easier to use than the above pair (plus more idiomatic in Java).

(btw, your signature is slightly wrong, it needs to be in a future)

**Comment by rkuhn** [comment]
Right you are—should we do this change? @pkinsky I think this would address your concern as well.

**Comment by ktoso** [comment]
I :+1: the IoResult :) (should be in FileSink/Source as well ofc)

**Comment by viktorklang** [comment]
This issue is mysteriously familiar. I believe I raised this earlier. But given GitHubs search capabilities I am not sure.

**Comment by pkinsky** [comment]
@rkuhn some kind of IoResult class seems like the correct solution. What do you think about also using it for InputStreamSource/OutputStreamSink, both of which materialize to a Future[Long] that completes with the number of bytes written/read?

**Comment by rkuhn** [comment]
Yes, indeed, such a class should be used for all IO endpoints.

**Comment by bchazalet** [comment]
For people trying to use the work around on version 2.0-M1, here is the updated solution:
``` scala
 def safeFileSink(f: File)(implicit ec: ExecutionContext): Sink[ByteString, Future[Long]] = {
    val fileSink = SynchronousFileSink(f)
    val errorMonitor = Sink.ignore
    Sink.fromGraph(FlowGraph.create(fileSink, errorMonitor)((f1, f2) => f2.flatMap(_ => f1)) { implicit builder =>
      (fileSink, errorMonitor) =>
        import FlowGraph.Implicits._

        val broadcast = builder.add(Broadcast[ByteString](2))

        broadcast.out(0) ~> fileSink
        broadcast.out(1) ~> errorMonitor

        SinkShape(broadcast.in)
    })
  }
```
Thanks @pkinsky for posting it in the first place, you saved my day!

**Comment by 2m** [comment]
Fixed in #19554.

**Comment by drewhk** [comment]
The backport of this is I think covered by: https://github.com/akka/akka/pull/20159/files
The IOResult change cannot be backported to 2.0.x as it is not binary compatible
